### PR TITLE
ROS2 port

### LINKED ISSUE #34: Porting to  ROS2 ??
There is some interest in porting this packages to ROS 2, see [this thread on Discours](https://discourse.ros.org/t/twist-mux-teleop-tools-porting-to-ros2/9448).

@bmagyar Could you please branch `dashing-devel` ??

### PR DESCRIPTION
**Pull request by artivis**
WIP PR for porting `teleop_tools` to ROS 2.

Main changes while porting:
- `key_teleop`: handle ctrl+c on terminal and window
- `joy_teleop`: action/service types have to be explicit in yaml config file, the functions to auto-deduce types from their topic aren't ported.

Fix #34 .

Port:
- [x] joy_teleop
- [x] key_teleop
- [x] mouse_teleop
- [x] teleop_tools_msgs

Let us add some unit-test alongside the port.

Tests:
- [x] linters / copyrights
- [ ] joy_teleop
- [ ] key_teleop
- [ ] mouse_teleop (no idea how to test this guy)

Review:
- [ ] license
- [ ] maintainers
- [ ] mails

### CODE REVIEW

**Comment by artivis** [comment]
During porting, I don't know if the `CHANGELOG` files should be kept are not. Similarly what about the package version ?? Any idea @bmagyar ?

**Comment by artivis** [comment]
`joy_teleop` is mostly ported, however I could only test the 'topic' actions with 'buttons'. It requires some more extensive testing before ticking the check-box.

**Comment by artivis** [comment]
@bmagyar @efernandez Could any of you tell me if the [`KeyTeleop`](https://github.com/ros-teleop/teleop_tools/pull/35/files#diff-9ef26010455755aae85cf154753be02aR121) class in `key_teleop` is useful for anything or can we simply remove it ??

Also I would propose to move the [`IncrementServer`](https://github.com/ros-teleop/teleop_tools/blob/kinetic-devel/joy_teleop/scripts/incrementer_server.py) somewhere else as it is not really related to `joy_teleop` and very much related to Tiago. What do you think ?
Maybe @v-lopez (Hi there:) ) has an idea about it ?

**Comment by artivis** [comment]
@bmagyar @efernandez This is a fairly large PR, let me know if you would like me to break it down into smaller pieces (e.g. PR per pkg).

**Comment by v-lopez** [comment]
> Also I would propose to move the [`IncrementServer`](https://github.com/ros-teleop/teleop_tools/blob/kinetic-devel/joy_teleop/scripts/incrementer_server.py) somewhere else as it is not really related to `joy_teleop` and very much related to Tiago. What do you think ?
> Maybe @v-lopez (Hi there:) ) has an idea about it ?

Hi!!
We rewrote the `IncrementerServer` in C++, in part to fix a drifting issue. [here](https://github.com/pal-robotics-forks/teleop_tools/commit/07f8c4e7227ca8f8f22fc0590f7447232de28105#diff-f0a3f4a2dbcd42c4ed8e3f69460a07bf)

I'd say it's useful to other robots, and its main (and so far) only user is `joy_teleop`. Although I agree that in itself is not related to it.

So I can't decide, having a separate package just for 120 lines of code seems overkill, but it does not entirely belong here.

**Comment by bmagyar** [comment]
I was thinking about the very same thing when I wrote the incrementerserver back in the day and decided to bundle it with joy_teleop as it is it's sole client and the amount of code does not justify making an entirely new packages for it. I'd argue it's not tiago-specific, it can be used for anything that requires incremental changes at the press of a button and has a position-like interface.

**Comment by bmagyar** [comment]
Package version: if ROS2 doesn't want to live in an entirely different universe as ROS, I suggest we bump the major version number. Considering that teleop_tools is quite mature, it's unlikely we'd want to bump major anytime soon in ROS.

**Comment by artivis** [comment]
Thanks @v-lopez @bmagyar for the feedback concerning the `IncrementServer`, I'll leave it their as part of the port :+1:

File: mouse_teleop/launch/mouse_teleop.launch.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+from ament_index_python.packages import get_package_share_directory
+from launch import LaunchDescription
+import launch_ros.actions
+
+
+def generate_launch_description():
+
+    parameters_file = os.path.join(
+        get_package_share_directory('mouse_teleop'),
```
[the remark this answers is not in the document]
Comment (artivis) [reply mouse_teleop/launch/mouse_teleop.launch.py]: It has not, see https://github.com/ros2/launch/issues/163.

File: mouse_teleop/package.xml [hunk 2]
```diff
@@ -1,17 +1,21 @@
 <?xml version="1.0"?>
-<package>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
```
[the remark this answers is not in the document]
Comment (artivis) [reply mouse_teleop/package.xml]: If I'm not mistaken this is used by the xml-linter test.

File: key_teleop/scripts/key_teleop.py [hunk 3]
```diff
@@ -75,129 +117,131 @@ def refresh(self):
     def beep(self):
         curses.flash()
 
-class KeyTeleop():
 
-    _interface = None
+# class KeyTeleop():
```
[the remark this answers is not in the document]
Comment (artivis) [reply key_teleop/scripts/key_teleop.py]: :+1:

File: key_teleop/config/key_teleop.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+key_teleop:
+  ros__parameters:
```
[the remark this answers is not in the document]
Comment (artivis) [reply key_teleop/config/key_teleop.yaml]: Yes. Odd isn't it ? xD

File: joy_teleop/test/test_copyright.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ament_copyright.main import main
+import pytest
+
+
+@pytest.mark.copyright
+@pytest.mark.linter
+def test_copyright():
```
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/test/test_copyright.py]: Unit testing the copyright header of the source files.

File: joy_teleop/package.xml [hunk 6]
```diff
@@ -1,20 +1,33 @@
 <?xml version="1.0"?>
-<package>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
```
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/package.xml]: If I'm not mistaken this is used by the xml-linter test.

File: joy_teleop/test/key_teleop.test.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+import os
+import time
+import unittest
+# import uuid
+
+import launch
+import launch_ros
+import launch_ros.actions
+# import launch_testing.util
```
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/test/key_teleop.test.py]: Comments in this file are there as a basis to further extend testing.

[the remark this answers is not in the document]
Comment (bmagyar) [reply joy_teleop/test/key_teleop.test.py]: please put those on a new WIP PR instead

File: joy_teleop/test/test_copyright.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply joy_teleop/test/test_copyright.py]: That's lovely but I'd be much happier to see more functional tests, not this sort of garbage (pardon my French). For ROS packages there is a top-level license tag that should cover every file that doesn't have specific files copyrighted.

File: mouse_teleop/package.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply mouse_teleop/package.xml]: You know what? Colcon, catkin, rosdep is going to error out when a linter would. I'd prefer to not lint the hell out of weightless stuff but focus on the important things, why don't add a Python linter instead?

File: joy_teleop/test/key_teleop.test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/test/key_teleop.test.py]: kk.

File: joy_teleop/test/test_copyright.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/test/test_copyright.py]: There is one functional test + those linters etc. That's as many more than before this port ;P

File: mouse_teleop/package.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (artivis) [reply mouse_teleop/package.xml]: There are `test/test_flake8.py` & `test/test_pop257.py` for python.

File: key_teleop/setup.cfg [hunk 8]
```diff
@@ -0,0 +1,4 @@
+[develop]
```
[the remark this answers is not in the document]
Comment (artivis) [reply key_teleop/setup.cfg]: ATM it is, I do not know how this can be baked back into `setup.py`.

**Comment by artivis** [comment]
@bmagyar would you give a quick look at authors/maintainers and especially their mails. Some are quite outdated.

**Comment by IanTheEngineer** [comment]
Just cross referencing here: I've opened a PR against @artivis 's fork for some Dashing fixes https://github.com/artivis/teleop_tools/pull/1

**Comment by artivis** [comment]
@IanTheEngineer thank you for those fixes ! It's true I couldn't test much the buttons-related parts as I don't have a remote atm and was too lazy to do it from the terminal :sweat_smile:. I'll merge your changes to this PR asap.

**Comment by artivis** [comment]
Merged https://github.com/artivis/teleop_tools/pull/1. Manually checked the whole example yaml :+1:

File: mouse_teleop/package.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply mouse_teleop/package.xml]: :+1:

File: joy_teleop/test/test_copyright.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply joy_teleop/test/test_copyright.py]: Absolutely true :D

File: key_teleop/setup.cfg (hunk 8 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply key_teleop/setup.cfg]: I strongly dislike this approach as it simply noise over content but let's go with it. I'll do some research later

File: joy_teleop/setup.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+from setuptools import setup
+
+package_name = 'joy_teleop'
+
+setup(
+    name=package_name,
+    version='0.3.0',
+    packages=find_packages(exclude=['test']),
+    data_files=[
+        ('share/' + package_name, ['package.xml']),
```
Comment (bmagyar) [inline joy_teleop/setup.py]: let's reduce copy pasta?

```
package_name = 'joy_teleop'
share_path = 'share/' + package_name

...

    data_files=[
        (share_path, ['package.xml']),
        (share_path + '/config/', ['config/joy_teleop_example.yaml']),
```

File: joy_teleop/test/joy_teleop.test.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+class TestJoyTeleop(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.context = rclpy.context.Context()
+        rclpy.init(context=cls.context)
+        cls.node = rclpy.create_node('test_node', context=cls.context)
+
+        # The demo node listener has no synchronization to indicate when it's ready to start
+        # receiving messages on the /chatter topic.  This plumb_listener method will attempt
```
Comment (bmagyar) [inline joy_teleop/test/joy_teleop.test.py]: ehem

File: joy_teleop/test/joy_teleop.test.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        executor = rclpy.executors.SingleThreadedExecutor(context=self.context)
+        executor.add_node(self.node)
+        try:
+            executor.spin_once(timeout_sec=timeout_sec)
+        finally:
+            executor.remove_node(self.node)
+
+    def test_topic_hello(self):
+
+        # Expect the talker to publish strings on '/chatter'
```
Comment (bmagyar) [inline joy_teleop/test/joy_teleop.test.py]: badly formatted Python copied over from tutorial

File: joy_teleop/test/joy_teleop.test.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        end_time = time.time() + 10
+        while time.time() < end_time:
+            self.publisher.publish(self.msg)
+            self.spin_rclpy(1.0)
+            if len(msgs_rx) > 2:
+                break
+
+        self.assertGreater(len(msgs_rx), 1)
+
+        for msg in msgs_rx:
```
Comment (bmagyar) [inline joy_teleop/test/joy_teleop.test.py]: First of all, `msgs_rx` is a terrible name in high level code. I also think this code is not really being used.

`cls.publisher` is never used, therefore no `joy_teleop` is being tested?

File: joy_teleop/test/test_key_teleop.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+import subprocess
+
+import ament_index_python
+
+import pytest
+
+
+testdata = glob.glob(
+    os.path.join(
+        ament_index_python.get_package_share_directory('joy_teleop'),
```
Comment (bmagyar) [inline joy_teleop/test/test_key_teleop.py]: sure you want `joy_teleop` here?
_and_ if no but it works, can we remove this? :D

File: key_teleop/package.xml [hunk 14]
```diff
@@ -1,16 +1,26 @@
...
 
-  <license>BSD</license>
+  <exec_depend>geometry_msgs</exec_depend>
+  <exec_depend>rclpy</exec_depend>
+
+  <test_depend>ament_copyright</test_depend>
+  <test_depend>ament_flake8</test_depend>
+  <test_depend>ament_pep257</test_depend>
+  <test_depend>python3-pytest</test_depend>
+  <!-- <test_depend>launch_ros</test_depend> -->
```
Comment (bmagyar) [inline key_teleop/package.xml]: commented stuff

File: key_teleop/scripts/key_teleop.py [hunk 15]
```diff
@@ -2,19 +2,57 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+#
 #
 # Authors:
 #   * Siegfried-A. Gevatter
+#   * Jeremie Deray (artivis)
 
 import curses
-import math
+# import math
```
[the remark this answers is not in the document]
Comment (bmagyar) [reply key_teleop/scripts/key_teleop.py]: yo

File: key_teleop/setup.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+from setuptools import setup
+
+package_name = 'key_teleop'
+
+setup(
+    name=package_name,
+    version='0.3.0',
+    packages=find_packages(exclude=['test']),
+    data_files=[
+        ('share/' + package_name + '/', ['package.xml']),
```
Comment (bmagyar) [inline key_teleop/setup.py]: if agreed to the above `shared_path` var, please do here too

File: key_teleop/test/test_xmllint.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (bmagyar) [inline key_teleop/test/test_xmllint.py]: is there a psosibility to share this garbage between the multiple packages in this repo or we have to maintain several copies?

**Comment by artivis** [comment]
@bmagyar One of the functionality of `joy_teleop` is to instante msg/srv/action given a name in the config file. Something similar is done [here](https://github.com/ros2/ros2cli/pull/301/files#diff-7657b90fe78a034436904c34b3d70c2bR153) (action only) using some helper class/functions from `rosidl*` (see [here](https://github.com/ros2/ros2cli/pull/301/files#diff-7657b90fe78a034436904c34b3d70c2bR25)). At the moment we do this 'manually' and I would propose to rely on those `rosidl*` thingy. The current implementation only support interfaces fully 'qualified' such as `std_msgs/msg/String` but not `std_msgs/String` while it should. Also we would avoid code duplication and benefit from further improvement from this core package. What do you think ?

File: key_teleop/test/test_xmllint.py (hunk 17 above)
Comment (artivis) [reply key_teleop/test/test_xmllint.py]: I'm not sure if tests are also tested actually. For other files, yup it is necessary.

File: joy_teleop/config/joy_teleop_example.yaml [hunk 18]
```diff
@@ -1,115 +1,115 @@
...
-  array:
-    type: topic
-    message_type: std_msgs/UInt8MultiArray
-    topic_name: bytes
-    deadman_buttons: [5]
-    message_value:
-      -
-        target: data
-        value: [1,3,3,7]
+    # stop_everything:
```
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/config/joy_teleop_example.yaml]: I have no idea of the type of this action. I would simply remove it.

File: joy_teleop/scripts/incrementer_server.py [hunk 19]
```diff
@@ -1,44 +1,121 @@
...
-                                                auto_start=False)
-        self._command_pub = rospy.Publisher("command",
-                                            JointTrajectory,
-                                            queue_size=1)
-        state = rospy.wait_for_message("state", JTCS)
+        super().__init__('incrementer_server')
+
+        self._has_new_message = False
+
+        self._as = ActionServer(self,
```
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/scripts/incrementer_server.py]: Made it a one-liner.

File: joy_teleop/config/joy_teleop_example.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply joy_teleop/config/joy_teleop_example.yaml]: I'd argue to have this, it is a useful example showing that multiple buttons can be specified for an action. Could you change the action to something dummy you know the type of?

**Comment by bmagyar** [comment]
Yes I agree adding the parsers seems to be a good idea. Does it support both the fully qualified and vague names?

**Comment by artivis** [comment]
> Does it support both the fully qualified and vague names?

It will, I'm waiting on https://github.com/ros2/rosidl_python/pull/78 which factorize the previously linked code to functions.

File: joy_teleop/config/joy_teleop_example.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (artivis) [reply joy_teleop/config/joy_teleop_example.yaml]: I removed this comment and made one of the subsequent example a multi-buttons one.

**Comment by artivis** [comment]
@bmagyar unlinke what I said before, the `rosidl*` functions are not doing much more than what 's already done here. We can either use them at the expense of another dependency arguing about being more future proof or leave it as it is for now and revisite later if necessary.

File: key_teleop/scripts/key_teleop.py [hunk 20]
```diff
@@ -252,13 +199,24 @@ def _publish(self):
...
+
+def main():
     try:
-        curses.wrapper(main)
-    except rospy.ROSInterruptException:
+        curses.wrapper(execute)
+    except KeyboardInterrupt:
         pass
+    # except curses.error as c:
+        # print(c)
```
Comment (bmagyar) [inline key_teleop/scripts/key_teleop.py]: more commented stuff

**Comment by bmagyar** [comment]
Thanks for sticking through the review, merging!

**Pull request merged**